### PR TITLE
Service upload 17 02

### PR DESCRIPTION
**Pull request by praveenKDass**
# Description
These recommendations are intended to promote code quality and team communication during software development. They cover a variety of topics, including ensuring that pull requests are submitted to the correct branch, documenting new methods, preserving consistency across many services, and avoiding typical blunders like accessing APIs or DB queries within loops. Sensitive data should not be uploaded, and changes to environment variables or database models should be executed consistently. Teams may work more effectively and develop higher-quality software by adhering to these standards.

## Type of change
Please choose appropriate options.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Enhancement (additive changes to improve performance)
- [ ]  This change requires a documentation update

## Checklist 

- [ ]  It's critical to avoid making needless file modifications in contributions, such as adding new lines, console logs, or additional spaces, to guarantee cleaner and more efficient code. Furthermore, eliminating unnecessary imports from a file might enhance code readability and efficiency.
- [ ]  Ensure that the pull request is assigned to the right base branch and that the development branch name contains the JIRA Task Id. Furthermore, each commit message should include the JIRA Task Id in the manner "ED-100: message".
- [ ]  Only update packages if it is mentioned and authorized in the design document, and make sure that you have the required permissions.
- [ ]  Avoid making API and database queries inside a loop as it can lead to performance issues and slow down the system.
- [ ]  When calling another function inside a given function, add comments explaining the purpose and meaning of the passed arguments and expected return values.
- [ ]  If adding a blank argument in a function, add a comment explaining the reason for the blank argument.
- [ ]  Before submitting a pull request, do a self-review of your code to ensure there are no conflicts with the base branch and all comments have been addressed.
- [ ]  Before merging a pull request, it's important to have other team members review it to catch any potential errors or issues
- [ ]  To maintain code integrity, it's important to remove all related changes when removing code during a code review.
- [ ]  If new constants, endpoints, or utility functions are introduced, it is important to check if they already exist in the service to avoid any duplication.
- [ ]  Whenever a new environment variable is added to a service, it's important to ensure that the necessary changes are made to related files such as ".env.sample" and "envVariables.js" to maintain consistency and avoid errors. Additionally, the new environment variable should be added to the devops repository to ensure that it is properly documented and accessible to the team.
- [ ]  When adding a new function to a service, it is important to document it with relevant information such as the name, parameters, and return value in a consistent format across all services. Additionally, if there are any changes to the API response, ensure that the documentation in the controllers is updated accordingly. 
- [ ]  Write a clear and concise commit message that describes the changes made.
- [ ]  Maintain consistent function signature and code across all services when adding a function to multiple services. Implement changes to database models in all services that use the same model. 
- [ ]  Use only let and const. Do not use var.
- [ ]  Make common functions for repetitive code blocks.
- [ ]  Avoid uploading sensitive information such as secret tokens or passwords in pull requests to ensure data security.
- [ ]  Maintain consistent indentation and spacing throughout the code.

### CODE REVIEW

File: controllers/v1/cloud-services/files.js [hunk 1]
```diff
@@ -79,110 +77,148 @@ module.exports = class Files {
...
+  async getDownloadableUrl(req) {
+    return new Promise(async (resolve, reject) => {
+      try {
+        let downloadableUrl = await filesHelpers.getDownloadableUrl(
+          req.body.filePaths
+        );
+
+        return resolve(downloadableUrl);
+      } catch (error) {
+        console.log(error);
```
Comment (ankitpws) [inline controllers/v1/cloud-services/files.js:156]: @praveenKDass Please remove Console log.

File: module/cloud-services/files/helper.js [hunk 2]
```diff
@@ -119,50 +143,125 @@ module.exports = class FilesHelper {
...
+
+  /**
+   * create public/asstes folder if its doesn't exits.
+   * @method
+   * @name createDirectoryForUploadFile
+   */
+
+  static createDirectoryForUploadFile(){
+   
+  if (!fs.existsSync(ROOT_PATH + "/public")) {
```
Comment (ankitpws) [inline module/cloud-services/files/helper.js:254]: why is this added here it should be global js file @praveenKDass 
as if condition only

File: generics/constants/common.js [hunk 3]
```diff
@@ -93,5 +93,6 @@
   DELETE_USER: "delete-user",
   USER: "User",
   USER_DELETE_MODULE: "userDelete",
-  OFF: "OFF"
+  OFF: "OFF",
+  UPLOAD_FILE:"v1/cloud-services/files/upload"
```
Comment (ankitpws) [inline generics/constants/common.js:97]: please add this for UPLOAD_FILE api/cloud-services/mlcore/v1/files/upload @praveenKDass

File: controllers/v1/cloud-services/files.js [hunk 4]
```diff
@@ -79,110 +77,148 @@ module.exports = class Files {
...
-
-            let signedUrl =
-            await filesHelpers.preSignedUrls(
-                 req.body.request,
-                 req.body.ref,
-                 req.userDetails ? req.userDetails.userId : ""
+      try {
+
+        const filename = path.basename(req.query.file);
+          if(filename === req.files.file.name){
```
Comment (ankitpws) [inline controllers/v1/cloud-services/files.js:187]: if its binary then don't validate just upload and if its form-data then upload.
and its not necessary that you get form data with the file key only. First check if any key is present and then read data from that key. @praveenKDass

**Review by ankitpws — CHANGES_REQUESTED** [review]
Reviewed

File: controllers/v1/cloud-services/files.js (hunk 4 above)
Comment (praveenKDass) [reply controllers/v1/cloud-services/files.js]: Resolved, Added condition for binary

File: generics/constants/common.js (hunk 3 above)
Comment (praveenKDass) [reply generics/constants/common.js]: Changed as api/cloud-services/mlcore/v1/files/upload

File: module/cloud-services/files/helper.js (hunk 2 above)
Comment (praveenKDass) [reply module/cloud-services/files/helper.js]: Moved to global.js file

File: controllers/v1/cloud-services/files.js (hunk 1 above)
Comment (praveenKDass) [reply controllers/v1/cloud-services/files.js]: removed console.log

File: controllers/v1/cloud-services/files.js [hunk 5]
```diff
@@ -79,110 +78,162 @@ module.exports = class Files {
...
+          const binaryData = Buffer.from(req.body);
+          await fs.promises.writeFile(localFilePath, binaryData);
+          uploadStatus = await filesHelpers.upload(localFilePath, req.query.file);
+
+          if(uploadStatus.status == 200){
+            return resolve({
+              status: httpStatusCode["ok"].status,
+            });
+          }
+        } else {
```
Comment (aks30) [inline controllers/v1/cloud-services/files.js:204]: @ankitpws Make this condition for mutlipart/form-data only.

File: module/files/helper.js [hunk 6]
```diff
@@ -172,20 +189,53 @@ module.exports = class FilesHelper {
...
+        });
       }
-    })
+    });
   }
+
+  /**
+   * Upload and get Download Url for a file.
+   * @method
+   * @name upload
```
Comment (aks30) [inline module/files/helper.js:209]: @ankitpws Please update the function signature.

File: module/files/helper.js [hunk 7]
```diff
@@ -112,55 +110,74 @@ module.exports = class FilesHelper {
...
+          if (!serviceUpload) {
+            response.url = await cloudClient.getSignedUrl(
+              bucket, // bucket name
+              file, // file path
+              linkExpireTime, // expire
+              actionPermission // read/write
+            );
+          } else {
+            response.url = `${process.env.PUBLIC_BASE_URL}/${constants.common.UPLOAD_FILE}?file=${file}`;
+            response.downloadableUrl = await cloudClient.getDownloadableUrl(
```
Comment (aks30) [inline module/files/helper.js:168]: @ankitpws If possible add it for all Providers.

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 19 Feb

File: controllers/v1/cloud-services/files.js (hunk 5 above)
Comment (ankitpws) [reply controllers/v1/cloud-services/files.js]: Done

File: module/files/helper.js (hunk 7 above)
Comment (ankitpws) [reply module/files/helper.js]: Added

File: module/files/helper.js (hunk 6 above)
Comment (ankitpws) [reply module/files/helper.js:208]: Done
